src/appleseed/renderer/modeling/aov/cryptomatte.cpp

```diff
@@ -0,0 +1,784 @@
+        double                          m_rcp_canvas_height;
+        AABB2u                          m_crop_window;
+        Image&                          m_aov_image;
+        int                             m_num_layers;
+        vector<WeightMap>&              m_pixel_samples;
+        map<pair<int, int>, NameMap>&   m_tile_name_map;
+        CryptomatteAOV::CryptomatteType m_layer_type;
+
+        // Copy of the FilteredTile::add().
+        void filter(
+            const float                 x,
+            const float                 y,
+            const AABB2u&               crop_window,
+            const uint32_t              m3hash)
+        {
+            assert(m_renderer_filter);
+
+            // Convert (x, y) from continuous image space to discrete image space.
+            const float dx = x - 0.5f;
+            const float dy = y - 0.5f;
+
+            // Find the pixels affected by this sample.
+            AABB2i footprint;
+            footprint.min.x = truncate<int>(fast_ceil(dx - m_renderer_filter->get_xradius()));
+            footprint.min.y = truncate<int>(fast_ceil(dy - m_renderer_filter->get_yradius()));
+            footprint.max.x = truncate<int>(fast_floor(dx + m_renderer_filter->get_xradius()));
+            footprint.max.y = truncate<int>(fast_floor(dy + m_renderer_filter->get_yradius()));
+
+            // Don't affect pixels outside the crop window.
+            footprint = AABB2i::intersect(footprint, crop_window);
+
+            // Bail out if the point does not fall inside the crop window.
+            // Only check the x coordinate; the y coordinate is checked in the loop below.
+            if (footprint.min.x > footprint.max.x)
+                return;
+
+            for (int ry = footprint.min.y; ry <= footprint.max.y; ++ry)
+            {
+                for (int rx = footprint.min.x; rx <= footprint.max.x; ++rx)
+                {
+                    const float weight = m_renderer_filter->evaluate(rx - dx, ry - dy);
+                    m_pixel_samples[(m_tile_origin_y + ry) * m_frame_width + (m_tile_origin_x + rx)][m3hash] += weight; // possible race conditions writing to pixels on the tile edges
+                }
+            }
+            return;
+        }
+    };
+
+    const char* CryptomatteObjectAOVModel = "cryptomatte_object_aov";
+    const char* CryptomatteMaterialAOVModel = "cryptomatte_material_aov";
+}
+
+
+//
+// Cryptomatte AOV class implementation.
+//
+
+struct CryptomatteAOV::Impl
+{
+    vector<WeightMap>                   m_pixel_samples;
+    map<pair<int, int>, NameMap>        m_tile_name_map;
+    unique_ptr<Image>                   m_image;
+    int                                 m_num_layers;
+    CryptomatteAOV::CryptomatteType     m_layer_type;
+};
+
+CryptomatteAOV::CryptomatteAOV(const ParamArray& params)
+  : AOV("cryptomatte", params)
+  , impl(new Impl())
+{
+    const string cryptomatte_type = params.get_optional<string>("cryptomatte_type", "object_names");
```

Comment:
You can pass a list of valid values to `get_optional()`, that way an error message will be emitted if the value is not valid.